README.md
# netramesh

![netramesh](media/logo.png)

[![CircleCI](https://circleci.com/gh/Lookyan/netramesh/tree/master.svg?style=svg)](https://circleci.com/gh/Lookyan/netramesh/tree/master)

Ultra light service mesh has main goals:

- high performance
- observability (Jaeger distributed tracing)
- simplicity of operation
- unlimited scalability
- any infrastructure compatibility
- transparency

Service mesh netramesh consists of two main parts:
- Transparent TCP proxy for microservices with original destination retrieval.
- Init container for network rules configuration (iptables based).

## Getting started

Check out [examples](./examples)

## Supported application level protocols
- HTTP/1.1 and lower

Also netra supports any TCP proto traffic (proxies it transparently).


## How it works

![main parts](media/netra_main_parts.png)

To intercept all TCP traffic netra uses [iptables redirect rules](./iptables-rules.sh). After applying them, TCP traffic goes firstly to netra sidecar. Netra sidecar determines original destination using SO_ORIGINAL_DST socket option. After that netra sidecar works in bidirectional stream processing mode and proxies all TCP packets through itself. If app level protocol is HTTP1, netra parses it and sends tracing span.

![traffic interception](media/netra_traffic_intercept.png)

## Injecting

For now netra supports only manual injecting.

## Basic configuration (environment variables)

### Netra init (network interception settings)

Env name| Description
---|---
NETRA_SIDECAR_PORT | netra sidecar listen port redirect to (defaults to 14956)
NETRA_SIDECAR_USER_ID | netra sidecar user id to avoid infinite redirect loops (defaults to 1337)
NETRA_SIDECAR_GROUP_ID | netra sidecar group id to avoid infinite redirect loops (defaults to 1337)
INBOUND_INTERCEPT_PORTS | inbound ports to intercept (defaults to *, all ports)
OUTBOUND_INTERCEPT_PORTS | outbound ports to intercept (defaults to *, all ports)
NETRA_INBOUND_PROBABILITY | inbound probability to route TCP sessions (defaults to 1)
NETRA_OUTBOUND_PROBABILITY | outbound probability to route TCP sessions (defaults to 1)


### Netra sidecar

Switches

Switch name| Description
---|---
--service-name| service name for jaeger distributed trace spans

Env name| Description
---|---
NETRA_LOGGER_LEVEL | logger level (defaults to info), supported values: debug, info, warning, error, fatal
NETRA_PORT | netra sidecar listen port (defaults to 14956)
NETRA_PPROF_PORT | netra sidecar pprof port (defaults to 14957)
NETRA_TRACING_CONTEXT_EXPIRATION_MILLISECONDS | tracing context mapping cache expiration (defaults to 5000)
NETRA_TRACING_CONTEXT_CLEANUP_INTERVAL | tracing context cleanup interval (defaults to 1000)
NETRA_HTTP_PORTS | comma separated ports to determine as HTTP1 protocol (no default)
NETRA_HTTP_REQUEST_ID_HEADER_NAME | header name to match inbound and outbound requests. Applications should propagate it (defaults to X-Request-Id)
HTTP_HEADER_TAG_MAP | comma separated HTTP header to jaeger span tag conversion (example: `x-session:http.session,x-mobile-info:http.x-mobile-info`)
HTTP_COOKIE_TAG_MAP | comma separated HTTP cookie value to span tag conversion (example: `sess:http.cookies.sess`)
NETRA_HTTP_X_SOURCE_HEADER_NAME | source HTTP header name. Automatically added to each outbound request in case this header absent in request (defaults to X-Source)
NETRA_HTTP_X_SOURCE_VALUE | source HTTP header value (defaults to netra)
NETRA_STATSD_ENABLED | enabling statsd. Set "true" to enable (defaults to false)
NETRA_STATSD_PREFIX | Statsd prefix for all metrics (defaults to "")
NETRA_STATSD_ADDRESS | Statsd gate (defaults to "")

Also it supports all env variables [jaeger go library](https://github.com/jaegertracing/jaeger-client-go#environment-variables) provides.

## Comparison with Istio and linkerd2

Why do we need one more service mesh solution? Istio and linkerd2 are perfect service mesh solutions with very powerful set of features. But unfortunately they add significant resource and performance overhead.
Netramesh main goal is providing observability to your distributed system with small overhead (approximately 10-50Mb on each netra sidecar) and 1ms of latency overhead. If you don't need entire set of features Istio and linkerd2 provide, but you need to collect distributed traces and obtain important information about your microservice interaction then netra is a great fit.


CHANGELOG.md
# 0.4.6
- X-Source netra value rewrites existing one

# 0.4.3
- Statsd support
- Inbound metrics sending

# 0.4.1
- Bugfix port (parent span resolving)

# 0.4
- Added remote_addr tag to Jaeger
- Added X-Source header to propagate origin service name which can be customized through `NETRA_HTTP_X_SOURCE_HEADER_NAME` and `NETRA_HTTP_X_SOURCE_VALUE` env variable

# 0.3.1
- HTTP parsing moved to vendored stdlib
- HTTP WriteBody optimized (CopyBuffer used instead of Copy)

# 0.3
- X-Request-Id exposed to ENV VAR

# 0.2.2
- Fixed HTTP HEAD check

# 0.2.1
- Fixed bug with HTTP HEAD responses with non zero Content-length header
- HTTP stdlib vendored

# 0.2
- Added probabilistic routing mechanism

# 0.1.3
- Avoid address allocations in connection handling

# 0.1.2
- Performance optimizations

# 0.1.1
- Improve performance of http handler. Avoid additional allocations.

# 0.1
- Logger, configuration and specific port traffic forwarding

# 0.0.7
- Added timeout tag to spans for client timeouts

# 0.0.6
- Added HTTP_HEADER_TAG_MAP and HTTP_COOKIE_TAG_MAP Env config vars for HTTP headers to tracing span tag conversion

# 0.0.5
- HTTP upgrade type fallback to TCP added

# 0.0.4
- Added TCP fallback when we can't handle http protocol parsing.

# 0.0.3
- Added X-Request-id header detection and tracing span context propagation.

# 0.0.2
- Fixed read deadlock when client doesn't close connection by itself. Now it supports HTTP/1.0 proto.

# 0.0.1
- Base functionality for tcp traffic proxy and HTTP/1.1 parsing. Jaeger tracing spans support. 


iptables-rules.sh
#!/usr/bin/env bash

NETRA_SIDECAR_PORT=${NETRA_SIDECAR_PORT:-14956}
NETRA_SIDECAR_USER_ID=${NETRA_SIDECAR_USER_ID:-1337}
NETRA_SIDECAR_GROUP_ID=${NETRA_SIDECAR_GROUP_ID:-1337}

INBOUND_INTERCEPT_PORTS=${INBOUND_INTERCEPT_PORTS:-*}
OUTBOUND_INTERCEPT_PORTS=${OUTBOUND_INTERCEPT_PORTS:-*}

NETRA_INBOUND_PROBABILITY=${NETRA_INBOUND_PROBABILITY:-1}
NETRA_OUTBOUND_PROBABILITY=${NETRA_OUTBOUND_PROBABILITY:-1}

function dump {
    iptables-save
}

trap dump EXIT

IFS=,

iptables -t nat -N NETRA_INBOUND
iptables -t nat -N NETRA_OUTBOUND

iptables -t nat -A PREROUTING \
    -m statistic --mode random --probability ${NETRA_INBOUND_PROBABILITY} \
    -j NETRA_INBOUND

iptables -t nat -A OUTPUT \
    -m statistic --mode random --probability ${NETRA_OUTBOUND_PROBABILITY} \
    -j NETRA_OUTBOUND

if [ "${INBOUND_INTERCEPT_PORTS}" == "*" ]; then
    iptables -t nat -A NETRA_INBOUND -p tcp -m tcp -j REDIRECT --to-ports ${NETRA_SIDECAR_PORT}
else
    for port in ${INBOUND_INTERCEPT_PORTS}; do
        iptables -t nat -A NETRA_INBOUND -p tcp -m tcp --dport ${port} -j REDIRECT --to-ports ${NETRA_SIDECAR_PORT}
    done
fi

# avoid loops
iptables -t nat -A NETRA_OUTBOUND -m owner --uid-owner ${NETRA_SIDECAR_USER_ID} -j RETURN
iptables -t nat -A NETRA_OUTBOUND -m owner --gid-owner ${NETRA_SIDECAR_GROUP_ID} -j RETURN
iptables -t nat -A NETRA_OUTBOUND -p tcp -o lo -d 127.0.0.1 -j RETURN

if [ "${OUTBOUND_INTERCEPT_PORTS}" == "*" ]; then
    iptables -t nat -A NETRA_OUTBOUND -p tcp -j REDIRECT --to-ports ${NETRA_SIDECAR_PORT}
else
    for port in ${OUTBOUND_INTERCEPT_PORTS}; do
        iptables -t nat -A NETRA_OUTBOUND -p tcp --dport ${port} -j REDIRECT --to-ports ${NETRA_SIDECAR_PORT}
    done
fi
